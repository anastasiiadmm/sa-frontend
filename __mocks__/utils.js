import * as reduxHooks from 'react-redux';

jest.mock('../src/redux/store');
jest.mock('react-redux');

export const mockedUseSelectors = jest.spyOn(reduxHooks, 'useSelector');
export const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');
