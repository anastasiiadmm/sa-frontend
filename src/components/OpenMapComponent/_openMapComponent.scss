.OpenMapComponent {
  position: relative;

  &__card-block {
    position: fixed;
    padding-left: 20px;
    top: 74px;
    z-index: 999;
    display: flex;
    min-width: 100%;
    flex-direction: row;
    justify-content: space-between;
    gap: 15px;
    align-items: center;
  }

  &__card-block-info {
    display: flex;
    flex-direction: row;
    gap: 15px;
    padding-right: 270px;
  }

  &__card-style {
    display: flex;
    right: 0;
    border-radius: 58px;
  }

  &__arrow-left {
    padding: 0 10px;
    cursor: pointer;
    width: 35px;
  }

  &__not_coordinates {
    height: 80vh;
    display: flex;
    justify-content: center;
    align-items: center;
    > h1 {
      font-weight: 400;
      font-size: 26px;
    }
  }

  &__header-title {
    display: flex;
    align-items: center;
    flex-wrap: wrap;
  }

  &__title {
    display: flex;
    margin: 0 !important;
    font-weight: 400 !important;
  }

  &__img-title {
    margin: 0 10px 0 22px;
    width: 28px;
  }

  &__subtitle {
    margin: 0 !important;
    padding-left: 7px;
    padding-right: 7px;
    font-weight: 700 !important;
    text-transform: uppercase;
    font-size: 20px;
    > span {
      display: block;
      font-size: 20px;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }
  }

  &__render_btn {
    position: absolute;
    bottom: 150px;
    background-color: #689f3a !important;
    color: #ffffff !important;
    font-size: 14px;
    right: 10px;
    width: 36px;
    height: 36px;
    z-index: 1000;
    padding: 0 !important;
  }

  &__loading {
    width: 100%;
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  &__title_maps {
    display: none;
  }

  &__title_uppercase {
    text-transform: uppercase;
  }

  .ant-card .ant-card-body {
    padding: 8px;
  }
}

.leaflet-left {
  right: 10px;
  bottom: 80px;
  left: auto;
  top: auto;
}

.leaflet-div-icon {
  background-color: transparent !important;
  border: none !important;
}

.img-tractor {
  width: 28px;
}

.ant-layout-content::-webkit-scrollbar {
  width: 0;
}

.ant-layout-content::-webkit-scrollbar {
  width: 0;
}

@media (max-width: 990px) {
  .OpenMapComponent {
    &__card-block-info {
      padding-right: 20px;
    }
    .leaflet-left {
      bottom: 140px;
    }


    &__render_btn {
      bottom: 210px;
    }

    &__title_maps {
      display: block;
    }

    &__header-title > button {
      position: relative;
      z-index: 100;
    }
  }

  .top_btn {
    width: 100%;
    position: absolute;
    top: -50px;
    left: 0;
    z-index: 100;
    box-shadow: 0 0 0 0 #ffffff!important;
    > div {
      width: 100%;
      padding: 0!important;
      > div > p {
        width: 100%;
        text-align: center;
        margin-top: -23px;
        font-weight: 500;
        font-size: 15px;
      }
    }
  }
}

@media (max-width: 720px) {
  .OpenMapComponent {
    &__card-block {
      top: auto;
      display: block;
      padding-left: 0;
    }
    &__card-block-info {
      width: 100%;
      justify-content: center;
      background-color: #ffffff;
      gap: 0;
      margin-top: -5px;
      padding-bottom: 10px;
    }

    &__card-style {
      box-shadow: 0 0 0 0 #ffffff!important;
    }

    &__subtitle {
      font-size: 14px!important;
      padding-left: 0;
      padding-right: 0;

      > span {
        font-size: 14px;
      }
    }

    &__img-title {
      margin: -3px 5px 0px -6px;
      width: 23px;
    }
  }

  .img-tractor {
    width: 23px;
  }

  .top_btn > div > div > p {
    border-bottom: 1px solid #EBF3FF;
    padding-bottom: 12px;
  }
}

@media (max-width: 540px) {
  .OpenMapComponent {

    &__card-block-info {
      width: 100%;
      justify-content: center;
      flex-wrap: wrap;
      background-color: #ffffff;
      gap: 0;
      margin-top: -5px;
      padding-bottom: 10px;
      padding-right: 0;
    }
  }
}
