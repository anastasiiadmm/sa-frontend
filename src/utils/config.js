export const apiURL = '';

export const defaultError = { detail: 'На сервере что-то пошло не так' };
