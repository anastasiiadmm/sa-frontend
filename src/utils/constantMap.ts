import L from 'leaflet';

import active from 'assets/images/icons/active_tracktor.svg';
import endB from 'assets/images/icons/endB.svg';
import inactive from 'assets/images/icons/inactive_tracktor.svg';
import startA from 'assets/images/icons/startA.svg';

export const duckIconStart = new L.Icon({
  iconUrl: startA,
  iconRetinaUrl: startA,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(30, 30),
  className: 'leaflet-div-icon',
});

export const duckIconEnd = new L.Icon({
  iconUrl: endB,
  iconRetinaUrl: endB,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(30, 30),
  className: 'leaflet-div-icon',
});
export const activeIcon = new L.Icon({
  iconUrl: active,
  iconRetinaUrl: active,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(28, 30),
  className: 'leaflet-div-icon',
});

export const inactiveIcon = new L.Icon({
  iconUrl: inactive,
  iconRetinaUrl: inactive,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(28, 30),
  className: 'leaflet-div-icon',
});
