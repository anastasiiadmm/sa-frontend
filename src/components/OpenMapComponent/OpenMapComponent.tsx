import { AimOutlined } from '@ant-design/icons';
import { Button, Card, Spin, Tooltip, Typography } from 'antd';
import bem from 'easy-bem';
import L, { LatLngExpression } from 'leaflet';
import React, { useEffect, useRef, useState } from 'react';
import { CircleMarker, MapContainer, Marker, Polyline, Popup, TileLayer } from 'react-leaflet';
import { useNavigate, useParams } from 'react-router';

import arrowLeft from 'assets/images/icons/arrow-left.svg';
import endB from 'assets/images/icons/endB.svg';
import locale from 'assets/images/icons/locale.svg';
import map from 'assets/images/icons/map.svg';
import startA from 'assets/images/icons/startA.svg';
import tractorBlue from 'assets/images/icons/tractor-blue.svg';
import Errors from 'components/Errors/Errors';
import { accountsSelector, fetchVehicleInfo } from 'redux/accounts/accountsSlice';
import { useAppDispatch, useAppSelector } from 'redux/hooks';
import { mapSelector, obtainingCoordinate } from 'redux/map/mapSlice';
import { socketApiSocket } from 'utils/config';
import 'components/OpenMapComponent/_openMapComponent.scss';

const { Title } = Typography;

const purpleOptions = { color: '#1358BF' };

const OpenMapComponent = () => {
  const b = bem('OpenMapComponent');
  const { id, field_name } = useParams();
  const { account } = useAppSelector(accountsSelector);
  const { vehicle, field } = useAppSelector(mapSelector);
  const markerRef = useRef(null);
  const [markerPosition, setMarkerPosition] = useState<[number, number] | null>(null);
  const [bounds] = useState<number[][]>([
    [-90, -180],
    [90, 180],
  ]);
  const dispatch = useAppDispatch();
  const history = useNavigate();
  const [socketMap, setSocketMap] = useState({
    status: '',
    latitude: '',
    longitude: '',
  });
  const { userVehicleInfo, userVehicleInfoLoading } = useAppSelector(accountsSelector);
  const [loadingMap, setLoadingMap] = useState(false);
  const [loadingMapUpdate, setLoadingMapUpdate] = useState(false);
  const [socketLoading, setSocketLoading] = useState(false);

  const moveSpeed = account?.coords_timeout ? account.coords_timeout * 1000 + 200 : 0;

  const connectWebSocket = (time: number) => {
    let connectionID = '';

    const connect = () => {
      const socket = new WebSocket(socketApiSocket);
      socket.onopen = () => {
        setSocketLoading(true);
        socket.send(
          JSON.stringify({
            kind: 'ping',
            timeout: time,
            vehicle_id: Number(id),
            connection_id: connectionID,
          }),
        );
      };

      if (socket.readyState === WebSocket.OPEN) {
        socket.send(
          JSON.stringify({
            kind: 'ping',
            timeout: time,
            vehicle_id: Number(id),
            connection_id: connectionID,
          }),
        );
      }

      socket.onmessage = (event) => {
        setSocketLoading(false);
        const messageData = JSON.parse(event.data);
        connectionID = messageData.connection_id;
        setSocketMap({ ...socketMap, status: messageData.kind });
        if (messageData?.data?.latitude && messageData?.data?.longitude) {
          setSocketMap({ ...socketMap, ...messageData.data });
        }
      };

      socket.onclose = (event) => {
        if (event.code === 1000) {
          socket.close();
        } else {
          connect();
        }
      };

      return socket;
    };

    const socket = connect();
    return socket;
  };

  useEffect(() => {
    if (account?.coords_timeout) {
      const socket = connectWebSocket(account?.coords_timeout);

      return () => {
        socket.close();
      };
    }
  }, [account?.coords_timeout]);

  useEffect(() => {
    dispatch(fetchVehicleInfo({ vehicleId: String(id), pageUrl: '1' }));
  }, [dispatch]);

  useEffect(() => {
    dispatch(
      obtainingCoordinate({
        id: Number(id),
        field_name: !field_name?.includes('local-tractor') ? `?job=${field_name}` : '',
      }),
    );

    if (field_name?.includes('local-tractor')) {
      setLoadingMap(true);
      setLoadingMapUpdate(true);
    }
  }, []);

  useEffect(() => {
    if (markerRef.current && socketMap.latitude && socketMap.longitude) {
      const newPosition = [Number(socketMap.latitude), Number(socketMap.longitude)];
      const start = markerPosition || newPosition;
      const end = newPosition;

      const distance = Math.sqrt((end[0] - start[0]) ** 2 + (end[1] - start[1]) ** 2);

      const duration = Math.max(moveSpeed, distance * 50);

      let progress = 0;
      const steps = 60;

      const timer = setInterval(() => {
        progress += 1 / steps;
        const currentLng = start[1] + (end[1] - start[1]) * progress;
        const currentLat = start[0] + (end[0] - start[0]) * progress;
        const currentPos: [number, number] = [currentLat, currentLng];
        setMarkerPosition(currentPos);

        if (progress >= 1) {
          clearInterval(timer);
        }
      }, duration / steps);

      return () => clearInterval(timer);
    }
  }, [socketMap.latitude, socketMap.longitude]);

  useEffect(() => {
    if (loadingMapUpdate) {
      setLoadingMapUpdate(false);
    }
  }, [loadingMapUpdate]);

  const duckIcon = new L.Icon({
    iconUrl: map,
    iconRetinaUrl: map,
    iconAnchor: new L.Point(14, 14),
    popupAnchor: new L.Point(14, 0),
    iconSize: new L.Point(28, 30),
    className: 'leaflet-div-icon',
  });

  const duckIconStart = new L.Icon({
    iconUrl: startA,
    iconRetinaUrl: startA,
    iconAnchor: new L.Point(14, 14),
    popupAnchor: new L.Point(14, 0),
    iconSize: new L.Point(30, 30),
    className: 'leaflet-div-icon',
  });

  const duckIconEnd = new L.Icon({
    iconUrl: endB,
    iconRetinaUrl: endB,
    iconAnchor: new L.Point(14, 14),
    popupAnchor: new L.Point(14, 0),
    iconSize: new L.Point(30, 30),
    className: 'leaflet-div-icon',
  });

  const centerMap = () => {
    if (socketMap.latitude && socketMap.longitude) {
      return [Number(socketMap.latitude), Number(socketMap.longitude)];
    }
    return [0, 0];
  };

  const centerMapHandler = () => {
    if (loadingMap) {
      return centerMap() as LatLngExpression;
    }

    if (field?.results.point_A_lon && field?.results.point_B_lat) {
      return [field?.results.point_A_lon, field?.results.point_A_lat] as LatLngExpression;
    }

    return centerMap() as LatLngExpression;
  };

  const renderHandler = () => {
    setLoadingMap(true);
    setLoadingMapUpdate(true);
  };

  const backHandler = () => {
    history(-1);
  };

  const positions = (): LatLngExpression[] => {
    return [
      [
        field?.results.point_A_lon as number,
        field?.results.point_A_lat as number,
      ] as LatLngExpression,
      [
        field?.results.point_B_lon as number,
        field?.results.point_B_lat as number,
      ] as LatLngExpression,
    ];
  };

  function getCoordinateByType(coordinates: LatLngExpression[], type: string): LatLngExpression {
    if (type === 'start') {
      return [field?.results.point_A_lon as number, field?.results.point_A_lat as number];
    }
    if (type === 'end') {
      return [field?.results.point_B_lon as number, field?.results.point_B_lat as number];
    }

    throw new Error('Invalid type provided');
  }

  const latLngBounds: L.LatLngBoundsExpression = L.latLngBounds(
    bounds.map((coords: number[]) => [coords[0], coords[1]]),
  );
  const findResults = vehicle?.results?.processing_data.find((item) => item.id === Number(id));

  // const lineMapHistory = () => { // доработаем, когда из api будут присылаться данные
  //   const width = Number(field.results?.tool_width);
  //   const halfWidth = width / 2;
  //   const a: LatLngExpression = [field.results.point_A_lon, field.results.point_A_lat];
  //   const b: LatLngExpression = [field.results.point_B_lon, field.results.point_B_lat];
  //
  //   const resA = field.results.point_A_lon + halfWidth * 0.000008983;
  //   const resB = field.results.point_B_lon + (halfWidth + width) * 0.000008983;
  //
  //   const bottomRight = [field.results.point_A_lat, resA];
  //   const topLeft = [field.results.point_B_lat, resB];
  //
  //   return [topLeft, bottomRight];
  // };

  if (
    vehicle?.loading ||
    field?.loading ||
    loadingMapUpdate ||
    socketLoading ||
    userVehicleInfoLoading
  ) {
    return (
      <div data-testid='loading' className={b('loading')}>
        <div>
          <Spin size='large' />
        </div>
      </div>
    );
  }

  if (vehicle?.errors?.detail || field?.errors?.detail) {
    return (
      <div>
        <Errors
          status={vehicle?.errors?.status || field?.errors?.status}
          detail={vehicle?.errors?.detail || field?.errors?.detail}
        />
      </div>
    );
  }

  const fieldOne: LatLngExpression[] | LatLngExpression[][] = [
    [
      [52.68477, 39.618794],
      [52.686644, 39.62108],
    ],
    [
      [52.684656, 39.619088],
      [52.688475, 39.623862],
    ],
    [
      [52.686795, 39.621271],
      [52.690209, 39.62566],
    ],
    [
      [52.688709, 39.624191],
      [52.690079, 39.625992],
    ],
    [
      [52.689934, 39.626383],
      [52.684515, 39.619354],
    ],
  ];

  const fieldTwo: LatLngExpression[] | LatLngExpression[][] = [
    [
      [52.703808, 39.625391],
      [52.707549, 39.610904],
    ],
    [
      [52.703903, 39.62546],
      [52.704985, 39.621277],
    ],
    [
      [52.705126, 39.620722],
      [52.707643, 39.61098],
    ],
    [
      [52.707799, 39.611086],
      [52.704948, 39.622093],
    ],
    [
      [52.70492, 39.622222],
      [52.704051, 39.625583],
    ],
    [
      [52.704229, 39.625725],
      [52.70639, 39.617251],
    ],
    [
      [52.706635, 39.616292],
      [52.707972, 39.61119],
    ],
  ];

  const tractorFieldThree: LatLngExpression[] | LatLngExpression[][] = [
    [
      [45.229611, 39.590714],
      [45.238571, 39.590604],
    ],
    [
      [45.23857, 39.590778],
      [45.237515, 39.590795],
    ],
    [
      [45.237444, 39.590787],
      [45.230102, 39.590888],
    ],
    [
      [45.238577, 39.590958],
      [45.229622, 39.591069],
    ],
  ];

  const generateUniqueId = (() => {
    let id = 0;
    return () => `rectangle-${id++}`;
  })();

  return (
    <div className={b()}>
      <div className={b('card-block')}>
        <Card className={b('card-style')} bordered={false}>
          <div className={b('header-title')}>
            <button type='button' className='btn_none_style' onClick={backHandler}>
              <img className={b('arrow-left')} src={arrowLeft} alt='arrow' />
            </button>
            <Title level={3} className={b('title')}>
              <img src={locale} alt='locale' className={b('img-title')} />
              <Tooltip
                title={findResults?.field_name}
                color='#BBBBBB'
                overlayInnerStyle={{ padding: '5px 15px', borderRadius: 15 }}
                placement='topLeft'
              >
                <p className={b('subtitle')}>
                  <span>{field.results?.task_UID}</span>
                </p>
              </Tooltip>
            </Title>
            <Title level={3} className={b('title')}>
              <img src={tractorBlue} alt='tractor' className={b('img-title img-tractor')} />
              <p className={b('subtitle')}>{userVehicleInfo?.vehicle?.description}</p>
            </Title>
          </div>
        </Card>
      </div>
      <Button onClick={() => renderHandler()} className={b('render_btn')}>
        <AimOutlined />
      </Button>
      <div className={b('map-block')}>
        {socketMap.latitude && socketMap.latitude ? (
          <MapContainer
            center={centerMapHandler()}
            zoom={27}
            minZoom={2}
            maxZoom={18}
            scrollWheelZoom
            maxBounds={latLngBounds}
            style={{ width: '100%', height: '100vh' }}
          >
            <TileLayer
              attribution='&copy; <a href="https://www.esri.com/">Esri</a>'
              url='https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}'
            />
            <CircleMarker
              ref={markerRef}
              center={markerPosition !== null ? markerPosition : (centerMap() as LatLngExpression)}
              opacity={0}
              radius={10}
            >
              <Marker
                position={
                  markerPosition !== null ? markerPosition : (centerMap() as LatLngExpression)
                }
                icon={duckIcon}
              >
                <Popup>
                  <span className={b('title_uppercase')}>
                    {userVehicleInfo?.vehicle?.description}
                  </span>
                </Popup>
              </Marker>
            </CircleMarker>

            <Polyline weight={3} pathOptions={purpleOptions} positions={positions()}>
              <Marker position={getCoordinateByType(positions(), 'start')} icon={duckIconStart}>
                <Popup>Start</Popup>
              </Marker>
              <Marker position={getCoordinateByType(positions(), 'end')} icon={duckIconEnd}>
                <Popup>End</Popup>
              </Marker>
            </Polyline>
            {field?.results?.point_A_lat === 39.6186 && field?.results?.point_A_lon === 52.684864
              ? fieldOne.map((bounds) => (
                  <Polyline
                    key={generateUniqueId()}
                    positions={bounds}
                    color='rgba(30, 191, 19, 0.5)'
                    weight={33}
                  />
                ))
              : field?.results?.point_A_lat === 39.625304 &&
                field?.results?.point_A_lon === 52.703667
              ? fieldTwo.map((bounds) => (
                  <Polyline
                    key={generateUniqueId()}
                    positions={bounds}
                    color='rgba(30, 191, 19, 0.5)'
                    weight={33}
                  />
                ))
              : tractorFieldThree.map((bounds) => (
                  <Polyline
                    key={generateUniqueId()}
                    positions={bounds}
                    color='rgba(30, 191, 19, 0.5)'
                    weight={33}
                  />
                ))}
          </MapContainer>
        ) : null}
        {socketMap.status === 'no_geo' ? (
          <div className={b('not_coordinates')}>
            <h1>Координаты не найдены</h1>
          </div>
        ) : null}
      </div>
    </div>
  );
};

export default OpenMapComponent;
