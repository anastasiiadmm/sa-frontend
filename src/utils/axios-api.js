import axios from "axios";
import {apiURL} from "utils/config";

export const axiosApi = axios.create({
  baseURL: apiURL,
});