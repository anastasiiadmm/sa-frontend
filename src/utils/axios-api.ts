import axios from 'axios';

import { checkForTokens, logoutUser } from 'redux/auth/authSlice';
import { apiURL } from 'utils/config';
import { logoutLocalStorage } from 'utils/token';
import store from 'redux/store';

const axiosApi = axios.create({
  baseURL: apiURL,
});

axiosApi.interceptors.request.use(async (config) => {
  const key = store.getState()?.auth?.tokens?.access;

  if (key) {
    config.headers = {
      ...config.headers,
      Authorization: `Bearer ${key}`,
      'Content-Type': 'application/json',
    } as any;
  }
  return config;
});

axiosApi.interceptors.response.use(
  async (config) => {
    return config;
  },
  async (error) => {
    const originalRequest = error.config;

    const statusCode = error?.response?.status;
    const index = store;
    const { tokens } = index.getState().auth;
    if (
      tokens &&
      statusCode === 401 &&
      error.config &&
      !error.config._isReady &&
      error.response.data.messages
    ) {
      originalRequest._isReady = true;
      try {
        const resp = await axiosApi.post('/accounts/refresh/', { refresh: tokens.refresh });
        if (resp.status === 200) {
          const newTokens = resp.data;
          axiosApi.defaults.headers.Authorization = `Bearer ${newTokens.access}`;
          const usersLocal = {
            user: index.getState()?.auth?.user,
            token: {
              access: resp.data.access,
              refresh: tokens.refresh,
            },
          };
          index.dispatch(checkForTokens(usersLocal));
          const obj: any = usersLocal;
          delete obj.token.refresh;
          localStorage.setItem('users', JSON.stringify(obj));
          window.dispatchEvent(new Event('storage'));
          return axiosApi(originalRequest);
        }
      } catch (e) {
        logoutLocalStorage();
        window.location.reload();
      }

      store.dispatch(logoutUser());
    }

    return Promise.reject(error);
  },
);

export default axiosApi;
