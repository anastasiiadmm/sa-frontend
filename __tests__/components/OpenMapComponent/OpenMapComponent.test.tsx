import "@testing-library/jest-dom";
import {mockedDispatch, mockedUseSelectors} from "../../../__mocks__/utils";
import {screen, render, cleanup} from "@testing-library/react";
import {BrowserRouter} from "react-router-dom";
import "../../../__mocks__/matchMedia.mock";
import OpenMapComponent from "../../../src/components/OpenMapComponent/OpenMapComponent";
import renderer from "react-test-renderer";


afterEach(cleanup);

jest.mock('react-leaflet', () => ({
    CircleMarker: jest.fn(),
    MapContainer: jest.fn(),
    Marker: jest.fn(),
    Polyline: jest.fn(),
    Popup: jest.fn(),
    TileLayer: jest.fn()
}));

const mapErrors = {
    vehicle: {
        results: {
            processing_data: []
        }
    },
    field: {
        results: [],
        errors: {
            detail: 'Ошибка на сервере'
        }
    }
}

const mapLoading = {
    vehicle: {
        results: {
            processing_data: []
        },
        loading: true
    },
    field: {
        results: [],
        errors: null
    }
}

const mapRender = {
    vehicle: {
        results: {
            "id": 60,
            "image": "/media/vehicles/default.jpg",
            "processing_data": [
                {
                    "field_name": "Unistrong_ST6E21300000413_20210101080037_Sn21414347",
                    "work_area": 1213.006,
                    "id": 1,
                    "attachments": {
                        "Mode": "AB直线",
                        "toolsWidth": "12",
                        "toolsName": "Борона 24 метра",
                        "skipOverlap": "-0.3",
                        "toolsWidthResult": "11.7",
                        "frontBack": "-2.5",
                        "leftRight": "0.0"
                    }
                }
            ],
            "vin_code": "tech-58",
            "code": "wdRg83n4",
            "state_number": "tech-58",
            "description": "tech-5123123",
            "first_name": "tech-58",
            "middle_name": "tech-58",
            "last_name": "tech-58",
            "last_latitude": "42.857614",
            "last_longitude": "74.599974",
            "enterprise": 1
        },
        loading: false
    },
    field: {
        results: [
            {
                "PointA": "42.858595,74.599831",
                "PointB": "42.857191,74.599775"
            },
            {
                "PointA": "42.857191,74.599775",
                "PointB": "42.857250,74.598730"
            },
            {
                "PointA": "42.857250,74.598730",
                "PointB": "42.857673,74.598749"
            }
        ],
        errors: null
    }
}

describe('OpenMapComponents', () => {
    test('errors',  () => {
        mockedUseSelectors.mockReturnValue(mapErrors);
        const dispatch = jest.fn();
        mockedDispatch.mockReturnValue(dispatch);

        render(
            <BrowserRouter>
                <OpenMapComponent />
            </BrowserRouter>
        );
        const errorsText = screen.queryByText('Ошибка на сервере');
        expect(errorsText).toBeInTheDocument();
    })

    test('loading',  () => {
        mockedUseSelectors.mockReturnValue(mapLoading);
        const dispatch = jest.fn();
        mockedDispatch.mockReturnValue(dispatch);

        render(
            <BrowserRouter>
                <OpenMapComponent />
            </BrowserRouter>
        );
        const loading = screen.queryByTestId('loading');
        expect(loading).toBeInTheDocument()
    })

    test('render', async () => {
        mockedUseSelectors.mockReturnValue(mapRender);
        const dispatch = jest.fn();
        mockedDispatch.mockReturnValue(dispatch);

        const tree = renderer
            .create(
                <BrowserRouter>
                    <OpenMapComponent />
                </BrowserRouter>,
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    })
})
