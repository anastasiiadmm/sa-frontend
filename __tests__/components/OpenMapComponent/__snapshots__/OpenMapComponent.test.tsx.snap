// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenMapComponents render 1`] = `
<div
  className="OpenMapComponent"
>
  <div
    className="OpenMapComponent__card-block"
  >
    <div>
      <div
        className="ant-card OpenMapComponent__card-style css-dev-only-do-not-override-15rg2km"
      >
        <div
          className="ant-card-body"
          style={{}}
        >
          <div
            className="OpenMapComponent__header-title"
          >
            <button
              className="btn_none_style"
              onClick={[Function]}
              type="button"
            >
              <img
                alt="arrow"
                className="OpenMapComponent__arrow-left"
                src={
                  {
                    "process": [Function],
                  }
                }
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="OpenMapComponent__card-block-info"
    >
      <div
        className="ant-card OpenMapComponent__card-style css-dev-only-do-not-override-15rg2km"
      >
        <div
          className="ant-card-body"
          style={{}}
        >
          <div
            className="OpenMapComponent__header-title"
          >
            <h3
              className="ant-typography OpenMapComponent__title css-dev-only-do-not-override-15rg2km"
              style={
                {
                  "WebkitLineClamp": undefined,
                }
              }
            >
              <img
                alt="locale"
                className="OpenMapComponent__img-title"
                src={
                  {
                    "process": [Function],
                  }
                }
              />
              <p
                className="OpenMapComponent__subtitle"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                <span />
              </p>
            </h3>
          </div>
        </div>
      </div>
      <div
        className="ant-card OpenMapComponent__card-style css-dev-only-do-not-override-15rg2km"
      >
        <div
          className="ant-card-body"
          style={{}}
        >
          <div
            className="OpenMapComponent__header-title"
          >
            <h3
              className="ant-typography OpenMapComponent__title css-dev-only-do-not-override-15rg2km"
              style={
                {
                  "WebkitLineClamp": undefined,
                }
              }
            >
              <img
                alt="tractor"
                className="OpenMapComponent__img-title img-tractor"
                src={
                  {
                    "process": [Function],
                  }
                }
              />
              <p
                className="OpenMapComponent__subtitle"
              />
            </h3>
          </div>
        </div>
      </div>
      <div
        className="ant-card OpenMapComponent__card-style css-dev-only-do-not-override-15rg2km"
      >
        <div
          className="ant-card-body"
          style={{}}
        >
          <div
            className="OpenMapComponent__header-title"
          >
            <h3
              className="ant-typography OpenMapComponent__title css-dev-only-do-not-override-15rg2km"
              style={
                {
                  "WebkitLineClamp": undefined,
                }
              }
            >
              <img
                alt="speedRound"
                className="OpenMapComponent__img-title img-tractor"
                src={
                  {
                    "process": [Function],
                  }
                }
              />
              <p
                className="OpenMapComponent__subtitle"
              >
                0
                 
                км/ч
              </p>
            </h3>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    className="ant-btn css-dev-only-do-not-override-15rg2km ant-btn-default OpenMapComponent__render_btn"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span
      aria-label="aim"
      className="anticon anticon-aim"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="aim"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <defs>
          <style />
        </defs>
        <path
          d="M952 474H829.8C812.5 327.6 696.4 211.5 550 194.2V72c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v122.2C327.6 211.5 211.5 327.6 194.2 474H72c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h122.2C211.5 696.4 327.6 812.5 474 829.8V952c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V829.8C696.4 812.5 812.5 696.4 829.8 550H952c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8zM512 756c-134.8 0-244-109.2-244-244s109.2-244 244-244 244 109.2 244 244-109.2 244-244 244z"
        />
        <path
          d="M512 392c-32.1 0-62.1 12.4-84.8 35.2-22.7 22.7-35.2 52.7-35.2 84.8s12.5 62.1 35.2 84.8C449.9 619.4 480 632 512 632s62.1-12.5 84.8-35.2C619.4 574.1 632 544 632 512s-12.5-62.1-35.2-84.8A118.57 118.57 0 00512 392z"
        />
      </svg>
    </span>
  </button>
  <div
    className="OpenMapComponent__map-block"
  />
</div>
`;

exports[`OpenMapComponents render technique map success 1`] = `
<div
  className="TechniqueMap"
>
  <div
    className="TechniqueMap__card-block"
  >
    <div>
      <div
        className="ant-card TechniqueMap__card-style css-dev-only-do-not-override-15rg2km"
      >
        <div
          className="ant-card-body"
          style={{}}
        >
          <div
            className="TechniqueMap__header-title"
          >
            <button
              className="btn_none_style"
              onClick={[Function]}
              type="button"
            >
              <img
                alt="arrow"
                className="TechniqueMap__arrow-left"
                src={
                  {
                    "process": [Function],
                  }
                }
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="TechniqueMap__card-block-info"
    >
      <div
        className="ant-card TechniqueMap__card-style css-dev-only-do-not-override-15rg2km"
      >
        <div
          className="ant-card-body"
          style={{}}
        >
          <div
            className="TechniqueMap__title"
          >
            <img
              alt="Всего техники"
              className="TechniqueMap__img-title"
              src={
                {
                  "process": [Function],
                }
              }
            />
            <div
              className="TechniqueMap__title-info"
            >
              <span
                className="ant-typography TechniqueMap__title css-dev-only-do-not-override-15rg2km"
                style={
                  {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                Всего техники
              </span>
              <span
                className="ant-typography css-dev-only-do-not-override-15rg2km"
                style={
                  {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                <strong>
                  0
                </strong>
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-card TechniqueMap__card-style css-dev-only-do-not-override-15rg2km"
      >
        <div
          className="ant-card-body"
          style={{}}
        >
          <div
            className="TechniqueMap__title"
          >
            <img
              alt="Активны"
              className="TechniqueMap__img-title"
              src={
                {
                  "process": [Function],
                }
              }
            />
            <div
              className="TechniqueMap__title-info"
            >
              <span
                className="ant-typography TechniqueMap__title css-dev-only-do-not-override-15rg2km"
                style={
                  {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                Активны
              </span>
              <span
                className="ant-typography css-dev-only-do-not-override-15rg2km"
                style={
                  {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                <strong>
                  0
                </strong>
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-card TechniqueMap__card-style css-dev-only-do-not-override-15rg2km"
      >
        <div
          className="ant-card-body"
          style={{}}
        >
          <div
            className="TechniqueMap__title"
          >
            <img
              alt="Неактивны"
              className="TechniqueMap__img-title"
              src={
                {
                  "process": [Function],
                }
              }
            />
            <div
              className="TechniqueMap__title-info"
            >
              <span
                className="ant-typography TechniqueMap__title css-dev-only-do-not-override-15rg2km"
                style={
                  {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                Неактивны
              </span>
              <span
                className="ant-typography css-dev-only-do-not-override-15rg2km"
                style={
                  {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                <strong>
                  0
                </strong>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="TechniqueMap__map-block"
  />
</div>
`;
